External System Outbound Endpoint

dm-ps opened the pull request.

dm-ps committed:
External System Outbound Endpoint

teosarca commented on backend/de.metas.externalsystem/src/main/java/de/metas/externalsystem/outboundendpoint/ExternalSystemOutboundEndpointRepository.java:
```diff
@@ -0,0 +1,61 @@
+	@NonNull
+	public ExternalSystemOutboundEndpoint getById(@NonNull final ExternalSystemOutboundEndpointId id)
+	{
+		final I_ExternalSystem_Outbound_Endpoint endpointRecord = InterfaceWrapperHelper.load(id, I_ExternalSystem_Outbound_Endpoint.class);
```
was caching considered?

teosarca commented on backend/de.metas.externalsystem/src/main/java/de/metas/externalsystem/scriptedexportconversion/ExternalSystemScriptedExportConversionService.java:
```diff
@@ -104,13 +110,22 @@ public Map<String, String> getParameters(
 			@NonNull final String outboundDataProcessRecordId)
 	{
+		final ExternalSystemOutboundEndpoint endpoint = externalSystemOutboundEndpointRepository.getById(config.getExternalSystemOutboundEndpointId());
+		final String outboundEndpointData;
```
refactor outboundEndpointData to a method called toJsonString(ExternalSystemOutboundEndpoint) and inline it to line 128: e.g.
parameters.put(PARAM_SCRIPTEDADAPTER_OUTBOUND_ENDPOINT_DATA, toJsonString(endpoint));

more, IMHO we should avoid using domain objects (e.g. ExternalSystemOutboundEndpoint) as DTOs.
I would suggest introducing JsonExternalSystemOutboundEndpoint.

teosarca commented on backend/de.metas.externalsystem/src/main/java/de/metas/externalsystem/scriptedexportconversion/ExternalSystemScriptedExportConversionService.java:
```diff
@@ -104,13 +110,22 @@ public Map<String, String> getParameters(
+		}
+		catch (final JsonProcessingException exception)
+		{
+			throw new RuntimeException("Failed converting endpoint's properties to string: " + endpoint, exception);
```
isn't AdempiereException available?

teosarca commented on backend/de.metas.externalsystem/src/main/java/de/metas/externalsystem/outboundendpoint/ExternalSystemOutboundEndpoint.java:
```diff
@@ -0,0 +1,64 @@
+
+@Builder
+@Value
+@Jacksonized
```
pls avoid using domain objects as DTOs

metas-ts replied:
agree. anyways we need a dedicated dto (in de.metas.common), so i can unmarshal it on the other side

teosarca reviewed (dismissed):
@dm-ps  @metas-ts  check my comments

mrie04 committed:
Replace removed app, api and edi java image (#21599)

* Replace removed app and api java image

* Replace removed edi java image

metas-ts committed:
Add support for temporary (OAuth-)tokens to camel-scriptedadapter

metasfresh
- Change ExternalSystem_Outbound_Endpoint.EMail to UserLogin
- Create DTO JsonExternalSystemOutboundEndpoint for ExternalSystemOutboundEndpoint
- Cache ExternalSystemOutboundEndpointRepository.getById
- Change parameter name constant PARAM_SCRIPTEDADAPTER_OUTBOUND_ENDPOINT_DATA to PARAM_SCRIPTEDADAPTER_OUTBOUND_ENDPOINT_PARAMETERS (for fear people would thing that it's about the actual data to be send)

camel-externalsystems
- support auth-type "oauth" by obtaining a temporary token

teosarca commented on backend/de.metas.externalsystem/src/main/java/de/metas/externalsystem/outboundendpoint/ExternalSystemOutboundEndpointRepository.java:
```diff
@@ -31,8 +33,18 @@
 	@NonNull
 	public ExternalSystemOutboundEndpoint getById(@NonNull final ExternalSystemOutboundEndpointId id)
+	{
+		return Check.assumeNotNull(
```
* rename getById0 to retrieveById
* make sure `retrieveById` is returning nonnull
* remote this Check.asusmeNotnNull

teosarca replied:
NOTE: AFAIR, if loader method returns null then that result won't be cached anyways...

teosarca committed:
QA OAuthAccessToken

dm-ps committed:
External System Outbound Endpoint small refactoring

teosarca commented on misc/services/camel/de-metas-camel-externalsystems/de-metas-camel-scriptedadapter/src/main/java/de/metas/camel/externalsystems/scriptedadapter/oauth/OAuthTokenManager.java:
```diff
@@ -0,0 +1,155 @@
+
+	@Value
+	@Builder
+	private static class CacheKey
```
pls delete this (now or in a next PR/issue)
